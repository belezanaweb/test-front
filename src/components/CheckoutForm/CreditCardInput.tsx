import React, { useState } from 'react'
import InputMask from 'react-input-mask'

interface Props {
  handleChange: Function
}

const CreditCardInput: React.FC<Props> = ({ handleChange }) => {
  const [value, setValue] = useState('')
  const [mask, setMask] = useState('')

  const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const digits = event.target.value
    setValue(digits)
    const newMask = /^3[47]/.test(digits) ? '9999.999999.99999' : '9999.9999.9999.9999'
    setMask(newMask)
    handleChange(event)
  }

  return (
    <InputMask
      required
      name="creditcard"
      placeholder="____.____.____.____"
      value={value}
      mask={mask}
      onChange={onChange}
    />
  )
}

export default CreditCardInput
