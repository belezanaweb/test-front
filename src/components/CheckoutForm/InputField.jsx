import React, { useContext } from 'react'
import CheckoutFormContext from '../../context/CheckoutFormContext'

const InputField = ({ name, label, children }) => {
  const formValues = useContext(CheckoutFormContext)
  const { isValid } = formValues[name] || {}
  return (
    <div className={`input-field ${!!formValues[name] && !isValid && 'input-field--error'}`}>
      <label htmlFor={name}>{label}</label>
      {children}
      {formValues[name] && !isValid && <span class="message-error">Campo inv√°lido</span>}
    </div>
  )
}

export default InputField
