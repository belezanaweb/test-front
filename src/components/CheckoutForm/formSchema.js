import * as yup from 'yup'

const formSchema = {
  creditcard: yup
    .mixed()
    .required()
    .test({
      message: 'Dados do cartão inválido',
      test: value => value && value.replace(/\./gi, '').replace(/_/gi, '').length >= 15
    }),
  cardholder: yup.string().required('Nome do titular é obrigatório'),
  expdate: yup
    .mixed()
    .required()
    .test({
      message: 'Data Inválida',
      test: value => value && value.replace(/\//gi, '').replace(/_/gi, '').length === 6
    }),
  cvv: yup
    .mixed()
    .required()
    .test({
      message: 'CVV Inválido',
      test: value => value && value.replace(/_/gi, '').length === 3
    })
}

export default formSchema
