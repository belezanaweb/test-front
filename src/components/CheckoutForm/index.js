export { default, checkoutScheme } from './CheckoutForm'
