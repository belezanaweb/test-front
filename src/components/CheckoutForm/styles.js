import styled from 'styled-components'
import { device } from '../../styles/breakpoints'

export const Form = styled.form`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  gap: var(--spacing-4);

  > div {
    width: 100%;
  }

  #expirationDate-group {
    width: calc(60% - 10px);
  }

  #verificationCode-group {
    width: calc(40% - 10px);
  }

  @media ${device.tablet} {
    #name-group {
      width: calc(50% - 10px);
    }
    #creditCardNumber-group {
      width: calc(50% - 10px);
    }
  }

  @media ${device.desktop} {
    #name-group {
      width: calc(100%);
    }
    #creditCardNumber-group {
      width: calc(100%);
    }
  }
`
