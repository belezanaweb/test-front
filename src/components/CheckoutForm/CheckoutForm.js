import React from 'react'
import PropTypes from 'prop-types'
import InputMask from 'react-input-mask'
import * as yup from 'yup'

import Card from '../Card'
import FormControl from '../FormControl'

import * as S from './styles'

export const checkoutScheme = yup.object().shape({
  name: yup.string().required('Campo obrigatório'),
  creditCardNumber: yup.string().max(20).required('Campo obrigatório'),
  expirationDate: yup
    .string()
    .required('Campo obrigatório')
    .matches(/([0-9]{2})\/([0-9]{4})/, 'Formato inválido. Exemplo: MM/AAAA'),
  verificationCode: yup
    .string()
    .required('Campo obrigatório')
    .matches(/([0-9]{3})/, 'Formato inválido. Exemplo: 123')
})

const CheckoutForm = ({ register, errors }) => (
  <Card title="CARTÃO DE CRÉDITO">
    <S.Form>
      <FormControl
        {...register('creditCardNumber')}
        label="Número de cartão:"
        placeholder="____.____.____.____"
        id="creditCardNumber"
        name="creditCardNumber"
        errorMessage={errors.creditCardNumber?.message}
        as={InputMask}
        mask="9999.9999.9999.9999"
      />
      <FormControl
        {...register('name')}
        label="Nome do Titular:"
        placeholder="Como no cartão"
        id="name"
        name="name"
        errorMessage={errors.name?.message}
      />
      <FormControl
        {...register('expirationDate')}
        label="Validade (mês/ano):"
        placeholder="__/____"
        id="expirationDate"
        name="expirationDate"
        errorMessage={errors.expirationDate?.message}
        as={InputMask}
        mask="99/9999"
      />
      <FormControl
        {...register('verificationCode')}
        label="CVV:"
        placeholder="___"
        id="verificationCode"
        name="verificationCode"
        errorMessage={errors.verificationCode?.message}
        as={InputMask}
        mask="999"
      />
    </S.Form>
  </Card>
)

CheckoutForm.propTypes = {
  errors: PropTypes.object,
  register: PropTypes.func.isRequired
}

CheckoutForm.propTypes = {
  errors: {}
}

CheckoutForm.defaultProps = {
  href: undefined
}

export default CheckoutForm
