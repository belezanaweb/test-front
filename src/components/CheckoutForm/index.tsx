import React, { useContext } from 'react'
import InputMask from 'react-input-mask'

import useForm from '../../helpers/useForm'
import CreditCardInput from './CreditCardInput'
import InputField from './InputField'
import CartSummary from '../CartSummary'
import CheckoutFormContext from '../../context/CheckoutFormContext'
import formSchema from './formSchema'
import { useHistory } from 'react-router-dom'
import { OrderContext } from '../../context/OrderContext'
import CartContext from '../../context/CartContext'
import Card from '../Card'
import { FormStyle } from './styles'
import { Button } from '../Button'

const CheckoutForm: React.FC = () => {
  const [, setOrder] = useContext(OrderContext)
  const cart = useContext(CartContext)
  const onSubmit = () => {
    // @ts-ignore: Object with any
    const haveErrors = Object.keys(values).filter(key => !values[key].isValid)
    if (haveErrors[0]) return
    sendDataToOrder()
  }
  const { values, handleChange, handleSubmit } = useForm(onSubmit, formSchema)
  const history = useHistory()

  const sendDataToOrder = () => {
    // @ts-ignore: Object with any
    const { creditcard, cardholder, expdate } = values
    const details = { creditcard, cardholder, expdate }
    // @ts-ignore: Object with any
    setOrder({
      cart,
      details
    })

    history.push('/pedido')
  }
  return (
    <FormStyle className="checkout-form">
      <CheckoutFormContext.Provider value={values}>
        <form onSubmit={handleSubmit}>
          <Card>
            <div className="row">
              <div className="col-xs-12">
                <InputField label="Número do Cartão" name="creditcard">
                  <CreditCardInput handleChange={handleChange} />
                </InputField>
              </div>
              <div className="col-xs-12">
                <InputField label="Nome do Titular" name="cardholder">
                  <input
                    required
                    onChange={handleChange}
                    name="cardholder"
                    placeholder="Como no cartão"
                  />
                </InputField>
              </div>
              <div className="col-xs-7">
                <InputField label="Validade (mês/ano)" name="expdate">
                  <InputMask
                    required
                    name="expdate"
                    placeholder="__/____"
                    mask="99/9999"
                    onChange={handleChange}
                  />
                </InputField>
              </div>
              <div className="col-xs-5">
                <InputField label="CVV" name="cvv">
                  <InputMask
                    required
                    name="cvv"
                    placeholder="___"
                    mask="999"
                    onChange={handleChange}
                  />
                </InputField>
              </div>
            </div>
          </Card>
          <CartSummary />
          <Button type="submit">Finalizar o pedido</Button>
        </form>
      </CheckoutFormContext.Provider>
    </FormStyle>
  )
}

export default CheckoutForm
