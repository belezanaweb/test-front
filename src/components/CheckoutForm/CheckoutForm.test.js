import React from 'react'
import { render } from '@testing-library/react'
import CheckoutForm from './CheckoutForm'

describe('CheckoutForm', () => {
  it('should render CheckoutForm', () => {
    const register = jest.fn()
    const errors = {}

    const { container } = render(<CheckoutForm register={register} errors={errors} />)

    container.firstChild

    expect(container.firstChild).toBeInTheDocument()
  })

  it('should match snapshot', () => {
    const register = jest.fn()
    const errors = {}

    const { container } = render(<CheckoutForm register={register} errors={errors} />)

    expect(container.firstChild).toMatchSnapshot()
  })
})
