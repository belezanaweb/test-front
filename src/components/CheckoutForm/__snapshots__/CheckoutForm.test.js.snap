// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutForm should match snapshot 1`] = `
<div>
  <h2
    class="sc-gtsrHT hIMsyH"
    data-testid="title"
  >
    CARTÃO DE CRÉDITO
  </h2>
  <div
    class="sc-bdnxRM fjlQvV"
  >
    <form
      class="sc-gKAaRy kEqFEf"
    >
      <div
        class="sc-dlnjwi koSwQA"
        id="creditCardNumber-group"
      >
        <label
          class="sc-eCApnc gvEJHa"
          data-testid="input-label"
          for="creditCardNumber"
        >
          Número de cartão:
        </label>
        <input
          aria-invalid="false"
          class="sc-hKFxyN foRTGB"
          id="creditCardNumber"
          name="creditCardNumber"
          placeholder="____.____.____.____"
        />
      </div>
      <div
        class="sc-dlnjwi koSwQA"
        id="name-group"
      >
        <label
          class="sc-eCApnc gvEJHa"
          data-testid="input-label"
          for="name"
        >
          Nome do Titular:
        </label>
        <input
          aria-invalid="false"
          class="sc-hKFxyN foRTGB"
          id="name"
          name="name"
          placeholder="Como no cartão"
        />
      </div>
      <div
        class="sc-dlnjwi koSwQA"
        id="expirationDate-group"
      >
        <label
          class="sc-eCApnc gvEJHa"
          data-testid="input-label"
          for="expirationDate"
        >
          Validade (mês/ano):
        </label>
        <input
          aria-invalid="false"
          class="sc-hKFxyN foRTGB"
          id="expirationDate"
          name="expirationDate"
          placeholder="__/____"
        />
      </div>
      <div
        class="sc-dlnjwi koSwQA"
        id="verificationCode-group"
      >
        <label
          class="sc-eCApnc gvEJHa"
          data-testid="input-label"
          for="verificationCode"
        >
          CVV:
        </label>
        <input
          aria-invalid="false"
          class="sc-hKFxyN foRTGB"
          id="verificationCode"
          name="verificationCode"
          placeholder="___"
        />
      </div>
    </form>
  </div>
</div>
`
